src/clp_ffi_js/ir/StreamReader.cpp

```diff
@@ -165,52 +189,76 @@ auto StreamReader::decode_range(size_t begin_idx, size_t end_idx) const -> Decod
         }
         message.append(parsed.value());
 
-        constexpr size_t cLogLevelPositionInMessages{1};
-        size_t log_level{cLogLevelNone};
-        // NOLINTNEXTLINE(readability-qualified-auto)
-        auto const log_level_name_it{std::find_if(
-                cLogLevelNames.begin() + cValidLogLevelsBeginIdx,
-                cLogLevelNames.end(),
-                [&](std::string_view level) {
-                    return message.substr(cLogLevelPositionInMessages).starts_with(level);
-                }
-        )};
-        if (log_level_name_it != cLogLevelNames.end()) {
-            log_level = std::distance(cLogLevelNames.begin(), log_level_name_it);
-        }
-
         m_ts_pattern.insert_formatted_timestamp(log_event.get_timestamp(), message);
 
         EM_ASM(
                 { Emval.toValue($0).push([UTF8ToString($1), $2, $3, $4]); },
                 results.as_handle(),
                 message.c_str(),
                 log_event.get_timestamp(),
-                log_level,
-                log_num
+                log_event.get_log_level(),
+                log_event_idx +1
         );
-        ++log_num;
+
     }
 
     return DecodedResultsTsType(results);
 }
 
+void StreamReader::filter_log_events(const emscripten::val& logLevelFilter) {
+    m_filtered_log_event_map.clear();
+
+    // Check if the filter is empty
+    if (logLevelFilter.isNull()) {
+        m_is_filtered = false;
+        return;
+    }
+
+    std::vector<int> filter_levels = emscripten::vecFromJSArray<int>(logLevelFilter);
+
+    // Filter log events based on the provided log levels
+    for (size_t index = 0; index < m_encoded_log_events.size(); ++index) {
+        const auto& logEvent = m_encoded_log_events[index];
+        if (std::find(filter_levels.begin(), filter_levels.end(), logEvent.get_log_level()) != filter_levels.end()) {
+            m_filtered_log_event_map.push_back(index);
+        }
+    }
+    m_is_filtered = true;
+}
+
+auto StreamReader::get_filtered_log_event_map() const ->  FilteredLogEventMapType {
+    if (false == m_is_filtered) {
+        return FilteredLogEventMapType(emscripten::val::null());
+    }
+
+    return FilteredLogEventMapType(emscripten::val::array(m_filtered_log_event_map));
+}
+
 StreamReader::StreamReader(
         StreamReaderDataContext<four_byte_encoded_variable_t>&& stream_reader_data_context
 )
         : m_stream_reader_data_context{std::make_unique<
                   StreamReaderDataContext<four_byte_encoded_variable_t>>(
                   std::move(stream_reader_data_context)
           )},
-          m_ts_pattern{m_stream_reader_data_context->get_deserializer().get_timestamp_pattern()} {}
+          m_ts_pattern{m_stream_reader_data_context->get_deserializer().get_timestamp_pattern()},
+          m_encoded_log_events(),
+          m_filtered_log_event_map(),
+          m_is_filtered(false) {}
```

Comment:
1. We do not need to init `m_encoded_log_events` and `m_filtered_log_event_map` given their default constructors will be invoked.
2. Let's use the brace syntax for inits.
```suggestion
          m_is_filtered{false} {}
```